bpo-34556: Add --upgrade-deps to venv module

cooperlees opened the pull request:
- This allows for pip + setuptools to be automatically upgraded to the latest version on PyPI
- Update documentation to represent this change

[bpo-34556](https://bugs.python.org/issue34556): Add --upgrade to venv module


<!-- issue-number: [bpo-34556](https://bugs.python.org/issue34556) -->
https://bugs.python.org/issue34556
<!-- /issue-number -->

cooperlees committed:
Add --upgrade-deps to venv module
- This allows for pip + setuptools to be automatically upgraded to the latest version on PyPI
- Update documentation to represent this change

bpo-34556: Add --upgrade to venv module

cooperlees committed:
Fix bin_dir to bin_path + fix unittest pathing

cooperlees committed:
Using logger.debug and not a new raw logging.debug in venv/__init__.py

matrixise commented on Doc/library/venv.rst:
```diff
@@ -123,12 +123,17 @@ creation according to their needs, the :class:`EnvBuilder` class.
       (defaults to ``None`` which means directory name of the environment would
       be used).
 
+    * ``upgrade_deps`` - Update the base venv modules to the latest on PyPI
```
here, you need to use two `-` and not one

cooperlees replied:
Thanks.

matrixise commented on Doc/using/venv-create.inc:
```diff
@@ -60,10 +60,18 @@ The command, if run with ``-h``, will show the available options::
+.. versionchanged:: 3.8
+   Add ``--upgrade-deps`` option to upgrade pip + setuptools to the latest on PyPI
+
+.. versionchanged:: 3.6
```
this change is out of the scope of this PR, there is another PR (https://github.com/python/cpython/commit/3921b1cc34c2fc8b8b480c19a95ec306de710fdd) 

Maybe you could create another PR for that point.

cooperlees replied:
Ahh ok. I just thought it wouldn't hurt to fix things. I will revert.

matrixise commented on Lib/venv/__init__.py:
```diff
@@ -250,7 +256,7 @@ def setup_python(self, context):
             if sysconfig.is_python_build(True):
                 # copy init.tcl
-                for root, dirs, files in os.walk(context.python_dir):
+                for root, _dirs, files in os.walk(context.python_dir):
```
why this change?

cooperlees replied:
dirs is not used in the loop, this is a flake recommended way to indicate that. Will revert so I get my feature landed :)

matrixise commented on Misc/NEWS.d/next/Core and Builtins/2019-05-05-18-09-40.bpo-34556.o9kfpu.rst:
```diff
@@ -0,0 +1 @@
+Add --upgrade-deps to venv module
```
Use the ReST markup like `--upgrade-deps`. Add "Patch by ..."

cooperlees replied:
Sure thing.

matrixise reviewed (changes requested):
Hi,

Thank you so much for your contribution but I am going to ask you to remove the unnecessary blank lines and formatting. We have to be focused on the main modification, your new feature and not on a new line or a new space in a comment. It's useful in some cases, but not really here. just go to the essential thing but for your new code, you can format it, no problem.

I hope you will understand.

Thank you again for your contrib, I think it could be really useful with `./python -m venv --upgrade-deps`

PS: I have not checked but you could add yourself into the `Misc/ACKS` file

cooperlees committed:
- Revert all formatting and lint changes
- Revert documentation fixes
- Add update_deps param to create() function
- Update NEWS file correctly

matrixise commented on Lib/venv/__init__.py:
```diff
@@ -363,13 +369,25 @@ def install_scripts(self, context, path):
 
+    def upgrade_dependencies(self, context):
+        logger.debug(
+            'Upgrading {} packages in {}'.format(CORE_VENV_DEPS, context.bin_path)
```
You could use the f-string here. it's python 3.8+

matrixise reviewed (changes requested):
small change with the f-string. thank you

cooperlees committed:
Move logging to f-string

cooperlees committed:
Merge pull request #1 from python/master

Sync with upstream cpython
